### PR TITLE
[HIVEMALL-132] Generalize f1score UDAF to support any Beta value

### PR DESCRIPTION
**Pull request by nzw0301**
## What changes were proposed in this pull request?

Make f1 function more general fmeasure function for any positive beta value.

## What type of PR is it?

Improvement

## What is the Jira issue?

[HIVEMALL-132](https://issues.apache.org/jira/browse/HIVEMALL-132)

## How was this patch tested?

Add `FMeasureUDAFTest`

## Checklist

(Please remove this section if not needed; check `x` for YES, blank for NO)

- [x] Did you apply source code formatter, i.e., `mvn formatter:format`, for your commit?

### CODE REVIEW

File: resources/ddl/define-all.hive [hunk 1]
```diff
@@ -543,8 +543,8 @@ create temporary function lr_datagen as 'hivemall.dataset.LogisticRegressionData
 -- Evaluating functions --
 --------------------------
 
-drop temporary function if exists f1score;
-create temporary function f1score as 'hivemall.evaluation.FMeasureUDAF';
+drop temporary function if exists fmeasure;
+create temporary function fmeasure as 'hivemall.evaluation.FMeasureUDAF';
```
Comment (myui) [inline resources/ddl/define-all.hive]: Could you remain alias for `f1score` in DDLs for backward compatibility.

```sql
-- alias for backward compatibility
drop temporary function if exists f1score;
create temporary function f1score as 'hivemall.evaluation.FMeasureUDAF';

drop temporary function if exists fmeasure;
...
```

**Comment by myui** [comment]
@nzw Could you update user guide to include `fmeasure` and `f1score` in `incubator-hivemall/docs/gitbook/eval/classification_measures.md` ?

`npm install gitbook-cli; gitbook install; gitbook serve` on docs/gitbook .

Also, could you revise the current Evaluation section of https://treasure-data.gyazo.com/5ec4b737dcedd55353f8126040ea5366 to

```
• Binary Classification metrics
  • Area Under the ROC Curve
• Regression metrics
• Ranking metrics
```

Refer examples in 
http://scikit-learn.org/stable/modules/classes.html#module-sklearn.metrics
https://turi.com/learn/userguide/evaluation/classification.html#f_scores

**Comment by myui** [comment]
Also, some other DDLs also needed to be updated. Please grep `tree_export` to know which DDLs to update.

**Comment by myui** [comment]
@nzw0301 Could you add test for binary (and for multi-label measure)?

**Comment by nzw0301** [comment]
@myui Update this PR

- Add UnitTests for binary and multi-label inputs
- Update DDL's files: Add `fmeasure` alias

I will update the documentation related to this PR.

**Comment by myui** [comment]
@nzw0301 when documentation is updated, could you remove `[WIP]` from PR title?

I'll review and merge then.

**Comment by nzw0301** [comment]
@myui Thanks, sure.

File: docs/gitbook/eval/multilabel_classification_measures.md [hunk 2]
```diff
@@ -0,0 +1,148 @@
...
+  select array("bird")               as actual, array("bird")        as predicted
+union all
+  select array("bird", "cat")        as actual, array("bird", "cat") as predicted
+union all
+  select array("cat", "dog", "bird") as actual, array("cat", "dog")  as predicted
+union all
+  select array("dog")                as actual, array("dog", "bird") as predicted
+)
+select
+  f1score(actual, predicted)
```
Comment (myui) [inline docs/gitbook/eval/multilabel_classification_measures.md]: Could you change the optional third argument to take `const string options` ?

`-beta 1.0 (default)`
`-average [micro (default), macro]`
http://scikit-learn.org/stable/modules/generated/sklearn.metrics.f1_score.html#sklearn.metrics.f1_score

`f1score(actual, predicted)` equals to `fmeasure(actual, predicted, '-beta 1.0 -average micro')` .

See `UDFWithOptions` and it's usage.

File: docs/gitbook/eval/multilabel_classification_measures.md [hunk 3]
```diff
@@ -0,0 +1,148 @@
...
+  select array("bird")               as actual, array("bird")        as predicted
+union all
+  select array("bird", "cat")        as actual, array("bird", "cat") as predicted
+union all
+  select array("cat", "dog", "bird") as actual, array("cat", "dog")  as predicted
+union all
+  select array("dog")                as actual, array("dog", "bird") as predicted
+)
+select
+  fmeasure(actual, predicted, 2)
```
Comment (myui) [inline docs/gitbook/eval/multilabel_classification_measures.md]: `fmeasure(actual, predicted, '-beta 2.0 -average macro')`

File: docs/gitbook/eval/multilabel_classification_measures.md (hunk 2 above)
Comment (nzw0301) [inline docs/gitbook/eval/multilabel_classification_measures.md]: Thank you for your review.
OK, I will update arguments.

**Comment by nzw0301** [comment]
@myui I update this PR. Could you review this PR code?

**Comment by myui** [comment]
Sure. 

@takuti Could you help reviewing this PR?

**Comment by takuti** [comment]
@myui Oh, sorry. I've noticed this just now. Sure, I'm going to review.

File: core/src/test/java/hivemall/evaluation/FMeasureUDAFTest.java [hunk 4]
```diff
@@ -0,0 +1,355 @@
...
+        double beta = 1.0;
+        String average = " micro";
+        setUpWithArguments(beta, average);
+
+        evaluator.init(GenericUDAFEvaluator.Mode.PARTIAL1, inputOIs);
+        evaluator.reset(agg);
+
+        evaluator.iterate(agg, new Object[] {actual, predicted});
+
+        Assert.assertEquals(0.5714285714285715d, agg.get(), 1e-5);
```
Comment (takuti) [inline core/src/test/java/hivemall/evaluation/FMeasureUDAFTest.java]: Could you write how did you get the expected result? (Spark similarly to `testMultiLabelF1MultiSamples`?)

**Comment by nzw0301** [comment]
@takuti Thank you for your useful review.
I will fix this PR based on your comment.

File: core/src/main/java/hivemall/evaluation/FMeasureUDAF.java [hunk 5]
```diff
@@ -18,118 +18,387 @@
...
+            if (average.equals("micro")) {
+                divisor = denom(tp, totalActual, totalPredicted, squareBeta);
+                numerator = (1.d + squareBeta) * tp;
+            } else if (average.equals("binary")) {
+                double precision = precision(tp, totalPredicted);
+                double recall = recall(tp, totalActual);
+                divisor = squareBeta * precision + recall;
+                numerator = (1.d + squareBeta) * precision * recall;
+            } else {
+                throw new UnsupportedOperationException();
```
Comment (nzw0301) [inline core/src/main/java/hivemall/evaluation/FMeasureUDAF.java]: Yes, you are right. Currently, `average` value is only `binary` or `micro`.
So, I remove this `else` statement and replace `else if ` to `else`.

File: resources/ddl/define-all.spark [hunk 6]
```diff
@@ -530,6 +530,9 @@ sqlContext.sql("CREATE TEMPORARY FUNCTION lr_datagen AS 'hivemall.dataset.Logist
 sqlContext.sql("DROP TEMPORARY FUNCTION IF EXISTS f1score")
 sqlContext.sql("CREATE TEMPORARY FUNCTION f1score AS 'hivemall.evaluation.FMeasureUDAF'")
```
Comment (myui) [inline resources/ddl/define-all.spark]: It is better to copy the old FMeasureUDAF.java as F1ScoreUDAF.java for a backward compatibility.

Then, `CREATE TEMPORARY FUNCTION f1score AS 'hivemall.evaluation.F1ScoreUDAF'`.

**Comment by myui** [comment]
@nzw0301 grep f1score in `resources/ddl`. It's not only for `define-all.spark`.

File: core/src/main/java/hivemall/evaluation/FMeasureUDAF.java [hunk 7]
```diff
@@ -18,118 +18,387 @@
...
+            }
+
             if (divisor > 0) {
-                return (2.d * precision * recall) / divisor;
+                return (numerator / divisor);
             } else {
-                return -1d;
+                return -1.d;
```
[the remark this answers is not in the document]
Comment (nzw0301) [reply core/src/main/java/hivemall/evaluation/FMeasureUDAF.java]: I agree with you. This value was based on previous [f1score code](https://github.com/apache/incubator-hivemall/blob/7205de1e959f0d9b96ac756e415d8a8ada7e92af/core/src/main/java/hivemall/evaluation/FMeasureUDAF.java#L121).
I'll change this value to `0.d` when `divisor` is zero.
And thank you for sharing good link.

File: core/src/main/java/hivemall/evaluation/FMeasureUDAF.java [hunk 8]
```diff
@@ -18,118 +18,387 @@
...
+            }
+        }
+
+        private static double denom(long tp, long totalActual, long totalPredicted,
+                double squareBeta) {
+            long lp = totalPredicted - tp;
+
+            if (lp < 0) {
```
[the remark this answers is not in the document]
Comment (nzw0301) [reply core/src/main/java/hivemall/evaluation/FMeasureUDAF.java]: No, always `total* - tp` is non negative.  I remove this if statement. Thanks.

File: docs/gitbook/eval/multilabel_classification_measures.md [hunk 9]
```diff
@@ -0,0 +1,144 @@
...
+  under the License.
+-->
+
+<!-- toc -->
+
+# Multi-label classification
+
+
+Multi-label classification problem is the task to predict the labels given categorized dataset.
+Each sample $$i$$ has $$l_i$$ labels, where $$L$$ is the number of unique labels in the dataset, and $$0 \leq  l_i \leq |L| $$.
```
[the remark this answers is not in the document]
Comment (nzw0301) [reply docs/gitbook/eval/multilabel_classification_measures.md]: Yes, I fixed it. Thanks!

File: docs/gitbook/eval/multilabel_classification_measures.md [hunk 10]
```diff
@@ -0,0 +1,144 @@
...
+## Micro F-measure
+
+
+F-measure is generalized F1-score and the weighted harmonic mean of recall and precision.
+
+The value is computed by the following equation:
+$$
+\mathrm{F}_{\beta} = (1+\beta^2) \frac
+{\sum_i |l_i \cap p_i |}
+{ \beta^2 (\sum_i |l_i \cap p_i | + \sum_i |p_i - l_i |) + \sum_i |l_i \cap p_i | + \sum_i |l_i - p_i |}
```
[the remark this answers is not in the document]
Comment (nzw0301) [reply docs/gitbook/eval/multilabel_classification_measures.md]: Thanks! it is wrong equation. (Since my test code is wrong order.)
I also changed `FMeasureAggregationBuffer.denom()` to understand easily.

**Comment by nzw0301** [comment]
@takuti @myui Thank you for your kind comments. 
I completed update based on reviews.

File: core/src/main/java/hivemall/evaluation/F1ScoreUDAF.java [hunk 11]
```diff
@@ -0,0 +1,134 @@
...
+                    if (actual.contains(p)) {
+                        countTp++;
+                    }
+                }
+                this.tp += countTp;
+                this.totalActual += numActual;
+                this.totalPredicted += numPredicted;
+            }
+
+            void merge(PartialResult other) {
```
Comment (takuti) [inline core/src/main/java/hivemall/evaluation/F1ScoreUDAF.java]: Oh, there is a bug here! 🐛 

Correct:

```java
void merge(PartialResult other) {
    this.tp += other.tp;
    this.totalActual += other.totalActual;
    this.totalPredicted += other.totalPredicted;
}
```

File: docs/gitbook/eval/binary_classification_measures.md [hunk 12]
```diff
@@ -0,0 +1,261 @@
...
+  select 0 as truth, 0 as predicted
+)
+select
+  fmeasure(truth, predicted, '-average micro')
+from data
+;
+
+-- 0.5;
+```
+
```
Comment (takuti) [inline docs/gitbook/eval/binary_classification_measures.md]: How about writing the difference between `f1score` and `fmeasure` here? It could be helpful to understand the concept of `-average micro` and replace the old `f1score` with `fmeasure`. For instance:

> It should be noted that, since the old `f1score(truth, predicted)` function simply counts the number of "matched" elements between `truth` and `predicted`, the above query is equivalent to:
```sql
WITH data as (
  select 1 as truth, 0 as predicted
union all
  select 0 as truth, 1 as predicted
union all
  select 0 as truth, 0 as predicted
union all
  select 1 as truth, 1 as predicted
union all
  select 0 as truth, 1 as predicted
union all
  select 0 as truth, 0 as predicted
)
select
  f1score(array(truth), array(predicted))
from data
;
```

File: core/src/test/java/hivemall/evaluation/FMeasureUDAFTest.java [hunk 13]
```diff
@@ -0,0 +1,393 @@
...
+        double beta = 1.0;
+        String average = " micro";
+        setUpWithArguments(beta, average);
+
+        evaluator.init(GenericUDAFEvaluator.Mode.PARTIAL1, inputOIs);
+        evaluator.reset(agg);
+
+        evaluator.iterate(agg, new Object[] {actual, predicted});
+
+        // TODO: describe the way to get this expected value by spark
```
[the remark this answers is not in the document]
Comment (nzw0301) [reply core/src/test/java/hivemall/evaluation/FMeasureUDAFTest.java]: Oh,  I forget to update this line. Thanks.

File: docs/gitbook/eval/binary_classification_measures.md (hunk 12 above)
Comment (nzw0301) [reply docs/gitbook/eval/binary_classification_measures.md]: Sounds good! thanks

**Comment by nzw0301** [comment]
@takuti Thank you for comments!
I run query above, I get the same result on movielens-1M.

**Comment by nzw0301** [comment]
I will check whether the return value is the same tomorrow.

**Comment by takuti** [comment]
Importantly, in case that the number of mappers is 1, fixing the bug in `merge()` does not change the output value; you might see the same result `0.42483920860540153` even if the bug exists.

Alternatively, let you test the following query (6 mappers will be shown for each `select` statement), and check if its output is same as `fmeasure(truth, predicted, '-average micro')`:

```sql
WITH data as (
  select 1 as truth, 0 as predicted
union all
  select 0 as truth, 1 as predicted
union all
  select 0 as truth, 0 as predicted
union all
  select 1 as truth, 1 as predicted
union all
  select 0 as truth, 1 as predicted
union all
  select 0 as truth, 0 as predicted
)
select
  f1score(array(truth), array(predicted))
from data
;
```

If the bug has been fixed correctly, output should be same as `fmeasure(truth, predicted, '-average micro) = 0.5`, while the buggy code returns `f1score(array(truth), array(predicted)) = 1.0`

File: core/src/main/java/hivemall/evaluation/F1ScoreUDAF.java (hunk 11 above)
Comment (myui) [reply core/src/main/java/hivemall/evaluation/F1ScoreUDAF.java]: oops. This should be fixed.

**Comment by nzw0301** [comment]
I tested @takuti's query. The buggy code (previous code) returns `1.0`.
On the other hand, the fixed code return correct value`0.5`.

**Comment by nzw0301** [comment]
But I found another issue for f1score and fmeasure.

Both function cannot work on `EMR v5.8.0`

```sql
hive> select f1score(array(1), array(1));
FAILED: IllegalArgumentException Size requested for unknown type: org.apache.hadoop.hive.ql.exec.UDAFEvaluator
```

```sql
hive> select fmeasure(array(1), array(1));
FAILED: IllegalArgumentException Size requested for unknown type: java.lang.String
```

However, they can work on `EMR v5.0.0`.
I don't know why the failures occur on newer EMR.

File: core/src/main/java/hivemall/evaluation/FMeasureUDAF.java [hunk 14]
```diff
@@ -18,118 +18,387 @@
...
+        @Override
+        public DoubleWritable terminate(@SuppressWarnings("deprecation") AggregationBuffer agg)
+                throws HiveException {
+            FMeasureAggregationBuffer myAggr = (FMeasureAggregationBuffer) agg;
+            double result = myAggr.get();
+            return new DoubleWritable(result);
+        }
+    }
+
+    public static class FMeasureAggregationBuffer extends
```
Comment (myui) [inline core/src/main/java/hivemall/evaluation/FMeasureUDAF.java]: `estimate` is required for estimating resulting size for `AbstractAggregationBuffer`.

Refer
https://github.com/apache/hive/blob/master/ql/src/java/org/apache/hadoop/hive/ql/udf/generic/GenericUDAFComputeStats.java#L195

**Comment by nzw0301** [comment]
The issue above is avoided by creating table:

```
create table data as (
  select 1 as truth, 1 as predicted
);
```

```
// ok
select
  fmeasure(array(truth), array(predicted))
from data
;

// ok
select
  f1score(array(truth), array(predicted))
from data
;
```

**Comment by myui** [comment]
It's Hive v2.2.0 bug.

Filed a ticket: https://issues.apache.org/jira/browse/HIVE-17406

**Comment by takuti** [comment]
@myui Would you double-check this? I can merge whenever you are ready.

**Comment by myui** [comment]
Let me see.

File: docs/gitbook/eval/auc.md [hunk 15]
```diff
@@ -100,7 +100,7 @@ Note that `floor(prob / 0.2)` means that the rows are distributed to 5 bins for
 
 # Difference between AUC and Logarithmic Loss
 
-Hivemall has another metric called [Logarithmic Loss](stat_eval.html#logarithmic-loss) for binary classification. Both AUC and Logarithmic Loss compute scores for probability-label pairs. 
+Hivemall has another metric called [Logarithmic Loss](stat_eval.html#logarithmic-loss) for binary classification. Both AUC and Logarithmic Loss compute scores for probability-label pairs.
```
Comment (myui) [inline docs/gitbook/eval/auc.md]: Missing link. `stat_eval.html` is deleted.

**Comment by myui** [comment]
@nzw0301 I'll fix and merge it. No need to update this PR.

**Comment by myui** [comment]
@nzw0301 LGTM 👍 Merged. Well done! (thank you for your review @takuti )

**Comment by nzw0301** [comment]
Thank you for your review and support @takuti @myui
